crates/bevy_ecs/src/schedule/stepping.rs

```diff
@@ -0,0 +1,1305 @@
+            // the skip list, but ignore any changes it makes to the cursor.
+            let (skip, _) = state.skipped_systems(schedule, 0, Action::Waiting);
+            (skip, Some(start))
+        };
+
+        // update the stepping frame cursor based on if there are any systems
+        // remaining to be run in the schedule
+        // Note: Don't try to detect the end of the render frame here using the
+        // schedule index.  We don't know all schedules have been added to the
+        // schedule_order, so only next_frame() knows its safe to reset the
+        // cursor.
+        match next_system {
+            Some(i) => self.cursor.system = i,
+            None => {
+                let index = schedule_index + 1;
+                println!(
+                    "Stepping::skipped_systems(): next schedule {} -> {}",
+                    schedule_index, index,
+                );
+                self.cursor = Cursor {
+                    schedule: index,
+                    system: self.first_system_index_for_schedule(index),
+                }
+            }
+        }
+
+        Some(list)
+    }
+}
+
+#[derive(Default)]
+struct ScheduleState {
+    /// per-system [`SystemBehavior`]
+    behaviors: HashMap<NodeId, SystemBehavior>,
+
+    /// order of NodeIds in the schedule
+    node_ids: Vec<NodeId>,
+
+    /// changes to system behavior that should be applied the next time
+    /// [`ScheduleState::progress()`] is called
+    updates: HashMap<TypeId, Option<SystemBehavior>>,
+
+    /// This field contains the first steppable system in the schedule.
+    first: Option<usize>,
+}
+
+impl ScheduleState {
+    // set the stepping behavior for a system in this schedule
+    fn set_behavior(&mut self, system: SystemIdentifier, behavior: SystemBehavior) {
+        self.first = None;
+        match system {
+            SystemIdentifier::Node(node_id) => {
+                self.behaviors.insert(node_id, behavior);
+            }
+            SystemIdentifier::Type(type_id) => {
+                self.updates.insert(type_id, Some(behavior));
+            }
+        }
+    }
+
+    // clear the stepping behavior for a system in this schedule
+    fn clear_behavior(&mut self, system: SystemIdentifier) {
+        self.first = None;
+        match system {
+            SystemIdentifier::Node(node_id) => {
+                self.behaviors.remove(&node_id);
+            }
+            SystemIdentifier::Type(type_id) => {
+                self.updates.insert(type_id, None);
+            }
+        }
+    }
+
+    // clear all system behaviors
+    fn clear_behaviors(&mut self) {
+        self.behaviors.clear();
+        self.first = None;
+    }
+
+    // apply system behavior updates by looking up the node id of the system in
+    // the schedule, and updating `systems`
+    fn apply_updates(&mut self, schedule: &Schedule) {
+        for (node_id, system) in schedule.systems().unwrap() {
```

Comment:
Add comment explaining that systems may be present multiple times in a schedule, so we iterate through all systems and lookup any changes for that system's behavior and apply them to the node.